Sources/DreamsNetworkInteraction.swift

```diff
@@ -0,0 +1,140 @@
+//
+//  DreamsNetworkInteraction
+//  Dreams
+//
+//  This Source Code Form is subject to the terms of the Mozilla Public
+//  License, v. 2.0. If a copy of the MPL was not distributed with this
+//  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
+//  Copyright © 2021 Dreams AB.
+//
+
+import Foundation
+import class WebKit.WKWebView
+
+// MARK: DreamsNetworkInteracting
+
+public protocol DreamsNetworkInteracting {
+    func didLoad()
+    func use(webView: WKWebView)
+    func use(delegate: DreamsDelegate)
+    func launch(with credentials: DreamsCredentials, locale: Locale)
+}
+
+extension DreamsNetworkInteracting {
+    func launch(with credentials: DreamsCredentials) {
+        launch(with: credentials, locale: Locale.current)
+    }
+}
+
+// MARK: DreamsNetworkInteraction
+
+public final class DreamsNetworkInteraction: DreamsNetworkInteracting {
+
+    private let webService: WebServiceType
+    private let configuration: DreamsConfiguration
+    
+    private weak var webView: WKWebView!
+    private weak var delegate: DreamsDelegate?
+    
+    init(configuration: DreamsConfiguration, webService: WebServiceType) {
+        self.configuration = configuration
+        self.webService = webService
+    }
+    
+    // MARK: Public
+    
+    public func didLoad() {
+        webService.delegate = self
+        setUpUserContentController()
+    }
+    
+    public func use(webView: WKWebView) {
+        self.webView = webView
+    }
+    
+    public func use(delegate: DreamsDelegate) {
+        self.delegate = delegate
+    }
+    
+    public func launch(with credentials: DreamsCredentials, locale: Locale) {
+        loadBaseURL(credentials: credentials, locale: locale)
+    }
+    
+    // MARK: Private
+
+    private func handle(event: ResponseEvent, with jsonObject: JSONObject?) {
+        switch event {
+        case .onIdTokenDidExpire:
+            guard let requestId = jsonObject?["requestId"] as? String else { return }
+            delegate?.handleDreamsCredentialsExpired { [weak self] credentials in
+                self?.update(idToken: credentials.idToken, requestId: requestId)
+            }
+        case .onTelemetryEvent:
+            guard let name = jsonObject?["name"] as? String,
+                  let payload = jsonObject?["payload"] as? JSONObject else { return }
+            delegate?.handleDreamsTelemetryEvent(name: name, payload: payload)
+        case .onAccountProvisionRequested:
+            guard let requestId = jsonObject?["requestId"] as? String else { return }
+            delegate?.handleDreamsAccountProvisioningInitiated { [weak self] in
+                self?.accountProvisionInitiated(requestId: requestId)
+            }
```

Comment:
Thank you, great question!

The pattern is functionally not that different from delegation (you are handed a reference to something and must call it when you're done) but the handling and completion is in the same place and that makes the code easier to read and understand.
The main difference is in a worst-case scenario where consumer is not using any async calls with passing closures but something else (delegation?), then he must keep the closure somewhere. This scenario may seem backwards (and it is cause this scenario is rather oldschool way of doing it) but the worst case is actually much cleaner cause consumer only knows what and how to call when its done, instead of previous approach where consumer: 
- knows about some ViewController/DreamsView existence 
- having concrete class handed in as a parameter (not clean at all) or kept in a closure/block
- keeping reference to a delegate and knowing more than is needed for account privision

Apple is using closures for networking and background app refresh that both can take a while or never complete.

Is my explanation convincing enough? Thanks for reading this 🙂